### ISSUE TITLE
Bug: [no-unnecessary-type-assertion] False positive in nullability flow analysis

### ISSUE THREAD

**Issue by Zarel** [typescript-eslint/typescript-eslint#10882]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.7.3&sourceType=script&fileType=.tsx&code=KYDwDg9gTgLgBAMwK4DsDGMCWEVxsAZxgAoA3AQwBslgAuOAIwgkuHJQEo4BvAKDjit4EJPAC8cFEkqUA3PziYEcMlRocFAkeLgBGAAz75ApSorVgGgQKjAYSKLm0BCAHQwIAZRhRMKAObEHPIAvrxAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false

### Repro Code

```TypeScript
export function test(value: boolean) {
  let out = null;
  if (value)
    out = 100;
  if (value)
    return out!.toString();
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-type-assertion": "error",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

No error

### Actual Result

```
  6:12  error  This assertion is unnecessary since it does not change the type of the expression  @typescript-eslint/no-unnecessary-type-assertion
```

### Additional Info

As you can see, removing the `!` causes a TypeScript error, so the assertion is in fact necessary. Adding a type annotation (`out: number | null = null`) makes the error go away.

**Comment by bradzacher** [comment]
Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/10334
See the TS issue filed for the TS bug: https://github.com/microsoft/TypeScript/issues/60514
